src/Features/Core/Portable/CodeLens/CodeLensFindReferenceProgress.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeLens
+{
+    /// <summary>
+    /// Tracks incremental progress of a find references search, we use this to
+    /// count the number of references up until a certain cap is reached and cancel the search
+    /// or until the search completes, if such a cap is not reached.
+    /// </summary>
+    /// <remarks>
+    /// All public methods of this type could be called from multiple threads.
+    /// </remarks>
+    internal sealed class CodeLensFindReferencesProgress : IFindReferencesProgress, IDisposable
+    {
+        /// <summary>
+        /// this token is linked to an aggregate token that is passed to the Find References
+        /// operation, this is used solely to trigger the cancellation of an in progress
+        /// find references operation, when the references count hits a given cap.
+        /// </summary>
+        private readonly CancellationTokenSource _cancellationTokenSource;
```

Comment:
I don't think you need two CancellationTokenSource. I think you just need _aggregateCancellationTokenSource.

just call cancel from _aggregateCancellationTokenSource.